File: shared/license-sdk/license_client.py
"""
─────────────────────────────────────────────────────────────────
SaaS Platform — License SDK para Python
Versão: 1.0.0

Use este arquivo no SEU SOFTWARE PYTHON para validar licenças.

Não requer dependências externas — usa apenas a stdlib do Python.

Uso:
    from license_client import LicenseClient

    client = LicenseClient('https://seudominio.com')
    result = client.validate('XXXX-XXXX-XXXX-XXXX')

    if not result['valid']:
        print(f"Licença inválida: {result['error']}")
        sys.exit(1)
─────────────────────────────────────────────────────────────────
"""

import hashlib
import json
import os
import platform
import socket
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path
from typing import Optional


class LicenseClient:
    """
    Cliente de validação de licenças para o SaaS Platform.

    Args:
        gateway_url:  URL base do gateway (ex: 'https://seudominio.com')
        cache_ttl:    Tempo em segundos que o cache offline é válido (padrão: 3600)
        offline_mode: Se True, usa cache quando servidor está offline (padrão: True)
        timeout:      Timeout em segundos para requisições (padrão: 8)
    """

    def __init__(
        self,
        gateway_url: str,
        cache_ttl: int = 3600,
        offline_mode: bool = True,
        timeout: int = 8,
    ):
        self.gateway_url  = gateway_url.rstrip("/")
        self.cache_ttl    = cache_ttl
        self.offline_mode = offline_mode
        self.timeout      = timeout
        self._cache_file  = Path.home() / ".saas_license_cache"
        self._id_file     = Path.home() / ".saas_machine_id"

    # ── Public API ─────────────────────────────────────────────────

    def validate(
        self,
        license_key: str,
        machine_name: Optional[str] = None,
    ) -> dict:
        """
        Valida uma chave de licença contra o servidor.

        Returns:
            dict com campos:
              valid (bool)     — se a licença é válida
              error (str)      — mensagem de erro se inválida
              code (str)       — código de erro
              license (dict)   — detalhes da licença (se válida)
              _from_cache (bool) — True se resultado veio do cache offline
        """
        machine_id   = self._get_machine_id()
        machine_name = machine_name or socket.gethostname()

        try:
            result = self._post("/api/licenses/validate", {
                "key":          license_key,
                "machine_id":   machine_id,
                "machine_name": machine_name,
            })

            if result.get("valid"):
                self._write_cache({
                    "license_key":  license_key,
                    "result":       result,
                    "validated_at": time.time(),
                })

            return result

        except (urllib.error.URLError, TimeoutError, OSError, ConnectionError) as e:
            if self.offline_mode:
                cached = self._read_cache()
                if cached and cached.get("license_key") == license_key:
                    age = time.time() - cached.get("validated_at", 0)
                    if age < self.cache_ttl:
                        return {
                            **cached["result"],
                            "_from_cache": True,
                            "_cache_age":  round(age),
                        }
                return {
                    "valid": False,
                    "error": "Servidor indisponível e cache expirado. Conecte-se à internet.",
                    "code":  "OFFLINE_CACHE_EXPIRED",
                }

            raise RuntimeError(f"Falha ao conectar ao servidor de licenças: {e}") from e

    def deactivate(self, license_key: str) -> bool:
        """Desativa esta máquina para a licença (antes de mover para outro PC)."""
        try:
            result = self._post("/api/licenses/deactivate", {
                "key":       license_key,
                "machine_id": self._get_machine_id(),
            })
            return result.get("success", False)
        except Exception:
            return False

    # ── Internal helpers ───────────────────────────────────────────

    def _post(self, path: str, body: dict) -> dict:
        url     = f"{self.gateway_url}{path}"
        payload = json.dumps(body).encode("utf-8")

        req = urllib.request.Request(
            url,
            data=payload,
            headers={
                "Content-Type": "application/json",
                "User-Agent":   "LicenseSDK-Python/1.0",
            },
            method="POST",
        )

        with urllib.request.urlopen(req, timeout=self.timeout) as res:
            return json.loads(res.read().decode("utf-8"))

    def _get_machine_id(self) -> str:
        if self._id_file.exists():
            return self._id_file.read_text().strip()

        # Build fingerprint from system info
        parts = [
            platform.node(),
            platform.system(),
            platform.machine(),
            platform.processor(),
            str(platform.python_version()),
        ]
        machine_id = hashlib.sha256("|".join(parts).encode()).hexdigest()[:32]
        self._id_file.write_text(machine_id)
        return machine_id

    def _write_cache(self, data: dict) -> None:
        try:
            self._cache_file.write_text(json.dumps(data))
        except Exception:
            pass

    def _read_cache(self) -> Optional[dict]:
        try:
            if self._cache_file.exists():
                return json.loads(self._cache_file.read_text())
        except Exception:
            pass
        return None


# ─────────────────────────────────────────────────────────────────
# Convenience function — validate and exit if invalid
# ─────────────────────────────────────────────────────────────────

def require_license(
    gateway_url: str,
    license_key: Optional[str] = None,
    env_var: str = "LICENSE_KEY",
    exit_on_fail: bool = True,
) -> dict:
    """
    Valida a licença e encerra o processo se inválida.

    Args:
        gateway_url:  URL do gateway
        license_key:  Chave de licença (ou None para ler de env_var)
        env_var:      Nome da variável de ambiente (padrão: LICENSE_KEY)
        exit_on_fail: Se True, chama sys.exit(1) se licença inválida

    Returns:
        dict com os detalhes da licença válida

    Example:
        license = require_license('https://seudominio.com')
        print(f"Bem-vindo, {license['customer']}!")
    """
    key = license_key or os.environ.get(env_var, "").strip()

    if not key:
        print(
            f"\n❌ Licença não encontrada.\n"
            f"   Defina {env_var}=XXXX-XXXX-XXXX-XXXX ou passe via argumento.\n"
            f"   Compre em: {gateway_url}/checkout.html\n"
        )
        if exit_on_fail:
            sys.exit(1)

    client = LicenseClient(gateway_url)
    result = client.validate(key)

    if not result["valid"]:
        error = result.get("error", "Desconhecido")
        code  = result.get("code", "")
        print(f"\n❌ Licença inválida: {error}")

        hints = {
            "PENDING_PAYMENT":        f"   Finalize o pagamento em: {gateway_url}/checkout.html",
            "MAX_ACTIVATIONS_REACHED": "   Limite de máquinas atingido. Contate o suporte.",
            "EXPIRED":                 f"   Renove sua licença em: {gateway_url}/checkout.html",
            "REVOKED":                 "   Esta licença foi revogada. Contate o suporte.",
            "NOT_FOUND":               "   Chave não encontrada. Verifique se digitou corretamente.",
        }
        if code in hints:
            print(hints[code])
        print()

        if exit_on_fail:
            sys.exit(1)

    lic = result.get("license", {})
    cached_notice = " (cache offline)" if result.get("_from_cache") else ""
    print(f"✅ Licença válida{cached_notice}")
    print(f"   Produto:   {lic.get('product', '-')}")
    print(f"   Cliente:   {lic.get('customer', '-')}")
    expires = lic.get("expires_at", None)
    print(f"   Expira em: {expires or 'Nunca (vitalícia)'}")
    print()

    return lic


# ─────────────────────────────────────────────────────────────────
# CLI — test from command line:
#   python license_client.py XXXX-XXXX-XXXX-XXXX https://seudominio.com
# ─────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Uso: python license_client.py <LICENSE_KEY> <GATEWAY_URL>")
        sys.exit(1)

    key = sys.argv[1]
    url = sys.argv[2]

    require_license(gateway_url=url, license_key=key)

File: shared/license-sdk/license-client.js
/**
 * ─────────────────────────────────────────────────────────────────
 * SaaS Platform — License SDK
 * Versão: 1.0.0
 *
 * Use este arquivo no SEU SOFTWARE para validar licenças em tempo real.
 *
 * Instalação:
 *   Copie este arquivo para o seu projeto ou instale via npm:
 *   npm install node-machine-id  (para obter o ID de máquina)
 *
 * Uso básico:
 *   const license = new LicenseClient('https://seudominio.com');
 *   const result  = await license.validate('XXXX-XXXX-XXXX-XXXX');
 *   if (!result.valid) { showError(result.error); process.exit(1); }
 * ─────────────────────────────────────────────────────────────────
 */

const https = require('https');
const http  = require('http');
const os    = require('os');
const crypto = require('crypto');
const fs    = require('fs');
const path  = require('path');

// ─────────────────────────────────────────────────────────────────
// LicenseClient
// ─────────────────────────────────────────────────────────────────
class LicenseClient {
  /**
   * @param {string} gatewayUrl - URL base do seu gateway (sem barra final)
   *   Exemplo: 'https://seudominio.com' ou 'http://localhost:3000'
   * @param {object} options
   * @param {string} options.cacheFile  - Caminho do cache local (padrão: .license_cache)
   * @param {number} options.cacheTTL   - TTL do cache em segundos (padrão: 3600 = 1h)
   * @param {boolean} options.offlineMode - Aceitar cache offline quando servidor indisponível
   */
  constructor(gatewayUrl, options = {}) {
    this.gatewayUrl  = gatewayUrl.replace(/\/$/, '');
    this.cacheFile   = options.cacheFile  || path.join(os.tmpdir(), '.saas_license_cache');
    this.cacheTTL    = options.cacheTTL   ?? 3600;
    this.offlineMode = options.offlineMode ?? true;
  }

  // ── Validate a license key ─────────────────────────────────────
  async validate(licenseKey, options = {}) {
    const machineId   = options.machineId   || this._getMachineId();
    const machineName = options.machineName || os.hostname();

    // 1. Try server validation
    try {
      const result = await this._request('POST', '/api/licenses/validate', {
        key: licenseKey,
        machine_id:   machineId,
        machine_name: machineName,
      });

      // Cache successful validation
      if (result.valid) {
        this._writeCache({ licenseKey, result, validatedAt: Date.now() });
      }

      return result;

    } catch (err) {
      // 2. If server unreachable + offline mode enabled → check cache
      if (this.offlineMode && this._isNetworkError(err)) {
        const cached = this._readCache();
        if (cached && cached.licenseKey === licenseKey) {
          const age = (Date.now() - cached.validatedAt) / 1000;
          if (age < this.cacheTTL) {
            return {
              ...cached.result,
              _fromCache: true,
              _cacheAge: Math.round(age),
            };
          }
        }
        return {
          valid: false,
          error: 'Servidor indisponível e cache expirado. Conecte-se à internet.',
          code: 'OFFLINE_CACHE_EXPIRED',
        };
      }
      throw err;
    }
  }

  // ── Check if license is still valid (lightweight poll) ────────
  async check(licenseKey) {
    return this.validate(licenseKey, { skipMachineCheck: true });
  }

  // ── Generate a machine fingerprint ────────────────────────────
  _getMachineId() {
    // Try to read a stored persistent ID first
    const idFile = path.join(os.tmpdir(), '.saas_machine_id');
    if (fs.existsSync(idFile)) {
      return fs.readFileSync(idFile, 'utf-8').trim();
    }

    // Generate from system info
    const info = [
      os.hostname(),
      os.platform(),
      os.arch(),
      os.cpus()?.[0]?.model || '',
      Object.values(os.networkInterfaces())
        .flat()
        .filter(n => !n.internal && n.mac !== '00:00:00:00:00:00')
        .map(n => n.mac)
        .sort()
        .join(','),
    ].join('|');

    const id = crypto.createHash('sha256').update(info).digest('hex').slice(0, 32);
    fs.writeFileSync(idFile, id, 'utf-8');
    return id;
  }

  // ── HTTP request helper ────────────────────────────────────────
  _request(method, path, body = null) {
    return new Promise((resolve, reject) => {
      const url      = new URL(this.gatewayUrl + path);
      const lib      = url.protocol === 'https:' ? https : http;
      const payload  = body ? JSON.stringify(body) : null;

      const options = {
        hostname: url.hostname,
        port:     url.port || (url.protocol === 'https:' ? 443 : 80),
        path:     url.pathname + url.search,
        method,
        headers: {
          'Content-Type':  'application/json',
          'User-Agent':    'LicenseSDK/1.0',
          ...(payload ? { 'Content-Length': Buffer.byteLength(payload) } : {}),
        },
        timeout: 8000,
      };

      const req = lib.request(options, (res) => {
        let data = '';
        res.on('data', chunk => (data += chunk));
        res.on('end', () => {
          try {
            resolve(JSON.parse(data));
          } catch {
            reject(new Error('Invalid JSON response from license server'));
          }
        });
      });

      req.on('error',   reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('NETWORK_TIMEOUT')); });

      if (payload) req.write(payload);
      req.end();
    });
  }

  _isNetworkError(err) {
    return ['ECONNREFUSED', 'ENOTFOUND', 'ETIMEDOUT', 'NETWORK_TIMEOUT'].some(
      c => err.message?.includes(c) || err.code === c
    );
  }

  _writeCache(data) {
    try {
      fs.writeFileSync(this.cacheFile, JSON.stringify(data), 'utf-8');
    } catch { /* ignore cache write errors */ }
  }

  _readCache() {
    try {
      if (!fs.existsSync(this.cacheFile)) return null;
      return JSON.parse(fs.readFileSync(this.cacheFile, 'utf-8'));
    } catch { return null; }
  }
}

// ─────────────────────────────────────────────────────────────────
// Python equivalent (copy to your Python project)
// ─────────────────────────────────────────────────────────────────
const PYTHON_SDK = `
# license_client.py — Python SDK
import hashlib, json, os, platform, socket, urllib.request, urllib.error, time

class LicenseClient:
    def __init__(self, gateway_url, cache_ttl=3600, offline_mode=True):
        self.gateway_url = gateway_url.rstrip('/')
        self.cache_ttl   = cache_ttl
        self.offline_mode = offline_mode
        self._cache_file = os.path.join(os.path.expanduser('~'), '.saas_license_cache')

    def validate(self, license_key, machine_name=None):
        machine_id   = self._get_machine_id()
        machine_name = machine_name or socket.gethostname()

        try:
            payload = json.dumps({
                'key': license_key,
                'machine_id': machine_id,
                'machine_name': machine_name,
            }).encode('utf-8')

            req = urllib.request.Request(
                f'{self.gateway_url}/api/licenses/validate',
                data=payload,
                headers={'Content-Type': 'application/json', 'User-Agent': 'LicenseSDK-Python/1.0'},
                method='POST'
            )
            with urllib.request.urlopen(req, timeout=8) as res:
                result = json.loads(res.read().decode())

            if result.get('valid'):
                self._write_cache({'license_key': license_key, 'result': result, 'validated_at': time.time()})
            return result

        except (urllib.error.URLError, TimeoutError, OSError):
            if self.offline_mode:
                cached = self._read_cache()
                if cached and cached.get('license_key') == license_key:
                    age = time.time() - cached.get('validated_at', 0)
                    if age < self.cache_ttl:
                        return {**cached['result'], '_from_cache': True}
            return {'valid': False, 'error': 'Servidor indisponível', 'code': 'OFFLINE'}

    def _get_machine_id(self):
        id_file = os.path.join(os.path.expanduser('~'), '.saas_machine_id')
        if os.path.exists(id_file):
            with open(id_file) as f: return f.read().strip()
        info = platform.node() + platform.system() + platform.machine() + platform.processor()
        machine_id = hashlib.sha256(info.encode()).hexdigest()[:32]
        with open(id_file, 'w') as f: f.write(machine_id)
        return machine_id

    def _write_cache(self, data):
        try:
            with open(self._cache_file, 'w') as f: json.dump(data, f)
        except: pass

    def _read_cache(self):
        try:
            if os.path.exists(self._cache_file):
                with open(self._cache_file) as f: return json.load(f)
        except: pass
        return None


# ── Usage example ──────────────────────────────────────────────
if __name__ == '__main__':
    client = LicenseClient('https://seudominio.com')
    result = client.validate('XXXX-XXXX-XXXX-XXXX')

    if not result['valid']:
        print(f"Licença inválida: {result['error']}")
        exit(1)

    print(f"✅ Licença válida — {result['license']['product']}")
    print(f"   Cliente:   {result['license']['customer']}")
    print(f"   Expira em: {result['license'].get('expires_at', 'Nunca (vitalícia)')}")
`;

// ─────────────────────────────────────────────────────────────────
// Usage examples
// ─────────────────────────────────────────────────────────────────
async function exampleUsage() {
  const client = new LicenseClient('https://seudominio.com', {
    offlineMode: true,
    cacheTTL: 3600, // 1 hour offline grace period
  });

  // ── Validate on startup ──────────────────────────────────────
  const licenseKey = process.env.LICENSE_KEY
    || readStoredKey()
    || promptUserForKey();

  console.log('Validando licença...');
  const result = await client.validate(licenseKey);

  if (!result.valid) {
    console.error(`\n❌ Licença inválida: ${result.error}`);
    console.error(`   Código: ${result.code}`);

    if (result.code === 'PENDING_PAYMENT') {
      console.error('   Acesse https://seudominio.com/app1/checkout para finalizar o pagamento.');
    } else if (result.code === 'MAX_ACTIVATIONS_REACHED') {
      console.error('   Limite de ativações atingido. Contate o suporte.');
    } else if (result.code === 'EXPIRED') {
      console.error('   Renove sua licença em https://seudominio.com/app1/checkout');
    }

    process.exit(1);
  }

  const lic = result.license;
  console.log(`✅ Licença válida${result._fromCache ? ' (cache offline)' : ''}`);
  console.log(`   Produto:   ${lic.product}`);
  console.log(`   Cliente:   ${lic.customer}`);
  console.log(`   Expira em: ${lic.expires_at || 'Nunca (vitalícia)'}`);
  console.log(`   Ativações: ${lic.activations_used}/${lic.activations_max}`);

  // ── Periodic re-validation (every 24h) ──────────────────────
  setInterval(async () => {
    const recheck = await client.validate(licenseKey);
    if (!recheck.valid) {
      console.error('Licença inválida. Encerrando...');
      process.exit(1);
    }
  }, 24 * 60 * 60 * 1000);
}

function readStoredKey() {
  const file = path.join(os.homedir(), '.myapp_license');
  return fs.existsSync(file) ? fs.readFileSync(file, 'utf-8').trim() : null;
}

function promptUserForKey() {
  // In a real app: show a dialog box
  return process.env.LICENSE_KEY || '';
}

module.exports = { LicenseClient };

// Uncomment to run example:
// exampleUsage().catch(console.error);
